src/crab/ebpf_domain.cpp

```diff
@@ -1227,9 +1459,13 @@ void ebpf_domain_t::operator()(const Bin& bin) {
                     },
                     [&](NumAbsDomain& inv) {
                         // Assertions should make sure we only perform this on non-shared pointers
-                        apply(inv, crab::arith_binop_t::SUB, dst.value, dst.offset, src.offset, true);
-                        inv -= dst.offset;
-                        inv -= dst.region_size;
+                        auto dst_offset = get_type_offset_variable(bin.dst, inv);
```

Comment:
This assumes a singleton type (as is checked on `get_type(inv, v)`). Otherwise we should do this for all valid offset variables

Reply:
Similar issue may happen on other locations in the code.

Reply:
In line 1459, same_type() can only return type for singleton types because they're intervals and same_type does subtraction and comparison against zero.  [a,b] - [a,b] is not zero, so this case will fall through to line 1482 and havoc the data.   We could do better by actually calculating an interval based on pointer deltas instead of going to top, but there's no actual bug here as far as I see.

Reply:
Updated code comments to clarify, and added a negative test case to single-instruction-binop.yaml to test this path.

Reply:
same_type() should be more precise than that, since the underlying numerical domain is capable of tracking equalities.

Reply:
I had considered that but had decided it was not safe.  Two registers both in type {packet,shared} are not known to have the same type at the point of comparison in the ebpf program.  One register could be packet and the other shared, so it was only safe to assume they're different and havoc the results.

Reply:
I'm not sure I understand.

If two registers may be of different types, then same_type will never return true. But it can correctly return true for two registers that are known to hold either a packet or shared. For example, immediately following a join of two branches where both had been assigned packet in one branch and shared in the other,

Reply:
The state is indistinguishable from a case where one branch set r1 to stack and r2 to packet, and another branch set r2 to stack and r1 to packet, and then the two branches joined.  They're different types, but the state is the same.

Reply:
The abstract state is only the same if the join operation of the Zones domain is imprecise and lost the equality information; in both branches it should have both t1-t2<=0 and t2-t1<=0, and therefore the result of the join should also have the same invariants.

Reply:
I didn't follow. Assuming by t1 and t2 you mean r1.type and r2.type, in my example r1.type and r2.type were never equal in any branch.  Can you elaborate?

Reply:
In your example they were not. In my example they were, and PREVAIL should be (and I believe is) able to distinguish these cases. The joined state should not be the same in the two examples.

Reply:
Example 1:
```
if *:
    <r1.type=stack, r2.type=stack, r1.type=r2.type>
else:
    <r1.type=packet, r2.type=packet, r1.type=r2.type>
<r1.type=stack|packet, r2.type=stack|packet, r1.type=r2.type>
```
Example 2:
```
if *:
    <r1.type=stack, r2.type=packet>
else:
    <r1.type=packet, r2.type=stack>
<r1.type=stack|packet, r2.type=stack|packet>
```

Reply:
Example 3:
```
if *:
    <r1.type=stack, r2.type=stack>
else:
    <r1.type=packet, r2.type=packet>
<r1.type=stack|packet, r2.type=stack|packet>
```

My point was that example 3 and example 2 are indistinguishable.  I think your point is that same_type(r1, r2) should look for an "r1.type=r2.type" invariant and if found, return true.  Correct?

Reply:
It should not matter whether or not they were independently set or not; since the information is available and expressible, the numerical domain should keep it. Example 2 is different from Example 3, but Example 3 and example 1 should have exactly the same result after the join, and exactly the same abstract states just before the join (not just equivalent - probably precisely the same representation, too). If you know that the current behavior is different, it is a bug.

`same_type()` checks whether `r1.type=r2.type` is entailed by the abstract state. It delegates the question to the underlying numerical domain, which is capable of precisely inferring such constraints by looking for two invariants - "r1.type-0<=r2.type" and "r2.type-0<=r1.type".

Reply:
Are you saying that in example 3, "r1.type=r2.type" should be explicitly added before the join?  My understanding is that would be another invariant added to the domain. 

Do you consider this a blocking issue or just a matter of improvement that should be done?

Reply:
I'm saying that the Zones domain will keep this invariant. The only difference between example 1 and 3 is whether you print it or not. 

Your choice does not lead to unsoundness, only to loss of precision. More precise would be to join over the subtraction relevant offsets of all possible pointer types in the intersection of r1.type and r2.type.
The soundness should come from a preceding `Assume` instruction, that should also consider equality of non-singleton invariants (which it currently does not: https://github.com/vbpf/ebpf-verifier/pull/289/files#diff-4260cca82870d77e078c224955ebce890cb127bdd92fd63a2d57e1374dedd9f2R645).

If you decide not to handle non-singleton invariants, please add a comment here and on `operator()(Assume)`, and change the comment above.

Reply:
It is not a blocking issue. I'm more concerned about the idea of making such deep changes to the underlying domain without any kind of mathematic proof, and without encapsulation (the new numerical domain is now baked into ebpf_domain_t), but I will let go. Hopefully I will have proofs and an implementation for a more uniform domain in the upcoming months.

Reply:
> If you decide not to handle non-singleton invariants, please add a comment here and on `operator()(Assume)`, and change the comment above.

It's not so much as deciding not to, it's not understanding how to do what you're suggesting.  I can update the comment if you'd like to take a crack at it subsequently.  Or if you want to educate me more and walk me through what would be needed I could try.

I agree proofs and a more uniform domain would be great in the future, though we need this now in the meantime.